files/blockinfile.py

```diff
@@ -0,0 +1,292 @@
+      - If specified, the block will be inserted before the last match of
+        specified regular expression. A special value is available; C(BOF) for
+        inserting the block at the beginning of the file.  If specified regular
+        expresion has no matches, the block will be inserted at the end of the
+        file.
+    choices: [ 'BOF', '*regex*' ]
+  create:
+    required: false
+    default: 'no'
+    choices: [ 'yes', 'no' ]
+    description:
+      - Create a new file if it doesn't exist.
+  backup:
+    required: false
+    default: 'no'
+    choices: [ 'yes', 'no' ]
+    description:
+      - Create a backup file including the timestamp information so you can
+        get the original file back if you somehow clobbered it incorrectly.
+"""
+
+EXAMPLES = r"""
+- name: insert/update "Match User" configuation block in /etc/ssh/sshd_config
+  blockinfile:
+    dest: /etc/ssh/sshd_config
+    block: |
+      Match User ansible-agent
+      PasswordAuthentication no
+
+- name: insert/update eth0 configuration stanza in /etc/network/interfaces
+        (it might be better to copy files into /etc/network/interfaces.d/)
+  blockinfile:
+    dest: /etc/network/interfaces
+    block: |
+      iface eth0 inet static
+          address 192.168.0.1
+          netmask 255.255.255.0
+
+- name: insert/update HTML surrounded by custom markers after <body> line
+  blockinfile:
+    dest: /var/www/html/index.html
+    marker: "<!-- {mark} ANSIBLE MANAGED BLOCK -->"
+    insertafter: "<body>"
+    content: |
+      <h1>Welcome to {{ansible_hostname}}</h1>
+      <p>Last updated on {{ansible_date_time.iso8601}}</p>
+
+- name: remove HTML as well as surrounding markers
+  blockinfile:
+    dest: /var/www/html/index.html
+    marker: "<!-- {mark} ANSIBLE MANAGED BLOCK -->"
+    content: ""
+"""
+
+
+def write_changes(module, contents, dest):
+
+    tmpfd, tmpfile = tempfile.mkstemp()
+    f = os.fdopen(tmpfd, 'wb')
+    f.write(contents)
+    f.close()
+
+    validate = module.params.get('validate', None)
+    valid = not validate
+    if validate:
+        if "%s" not in validate:
+            module.fail_json(msg="validate must contain %%s: %s" % (validate))
+        (rc, out, err) = module.run_command(validate % tmpfile)
+        valid = rc == 0
+        if rc != 0:
+            module.fail_json(msg='failed to validate: '
+                                 'rc:%s error:%s' % (rc, err))
+    if valid:
+        module.atomic_move(tmpfile, dest)
+
+
+def check_file_attrs(module, changed, message):
+
+    file_args = module.load_file_common_arguments(module.params)
+    if module.set_file_attributes_if_different(file_args, False):
+
+        if changed:
+            message += " and "
+        changed = True
+        message += "ownership, perms or SE linux context changed"
+
+    return message, changed
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            dest=dict(required=True, aliases=['name', 'destfile']),
```

Comment:
2.0 has a new type='path' that will expand user and expand vars for that parameter.